tests/e2e/steps_start_chains.go

```diff
@@ -29,6 +32,144 @@ func stepStartProviderChain() []Step {
+	s := []Step{
+		{
+			Action: CreateConsumerChainAction{
+				Chain:         ChainID("provi"),
+				From:          ValidatorID("alice"),
+				ConsumerChain: ChainID(consumerName),
+				SpawnTime:     uint(time.Minute * 3),
+				InitialHeight: clienttypes.Height{RevisionNumber: 0, RevisionHeight: 1},
+				TopN:          0,
+			},
+			State: State{
+				ChainID("provi"): e2e.ChainState{
+					ProposedConsumerChains: &proposedChains,
+				},
+			},
+		},
+	}
+
+	// Assign validator keys
+	// add a consumer key before the chain starts
+	// the key will be present in the consumer genesis initial_val_set
+	for _, valId := range validators {
+		valCfg := getDefaultValidators()[valId]
+		// no consumer-key assignment needed for validators using provider's public key
+		if !valCfg.UseConsumerKey {
+			continue
+		}
+		step := Step{
+			Action: AssignConsumerPubKeyAction{
+				Chain:          ChainID(consumerName),
+				Validator:      valId,
+				ConsumerPubkey: valCfg.ConsumerValPubKey,
+				// consumer chain has not started
+				// we don't need to reconfigure the node
+				// since it will start with consumer key
+				ReconfigureNode: false,
+			},
+			State: State{
+				ChainID(consumerName): ChainState{
+					AssignedKeys: &map[ValidatorID]string{
+						valId: valCfg.ConsumerValconsAddressOnProvider,
+					},
+					ProviderKeys: &map[ValidatorID]string{
+						valId: valCfg.ValconsAddress,
+					},
+				},
+			},
+		}
+		s = append(s, step)
+	}
+
+	// Opt-in Validators
+	for _, valId := range validators {
+		step := Step{
+			Action: OptInAction{
+				Chain:     ChainID(consumerName),
+				Validator: valId,
+			},
+			State: State{},
+		}
+		s = append(s, step)
+	}
+
+	// Launch chain
+	step := Step{
+		Action: UpdateConsumerChainAction{
+			Chain:         ChainID("provi"),
+			From:          ValidatorID("alice"),
+			ConsumerChain: ChainID(consumerName),
+			SpawnTime:     0, // launch now
+			InitialHeight: clienttypes.Height{RevisionNumber: 0, RevisionHeight: 1},
+			TopN:          0,
+		},
+		State: State{},
+	}
+	s = append(s, step)
+
+	// Setup validators for  chain
+	startChainVals := []StartChainValidator{}
+	valBalance := map[ValidatorID]uint{
+		ValidatorID("alice"): 0,
+		ValidatorID("bob"):   0,
+		ValidatorID("carol"): 0,
+	}
+
+	for idx, val := range validators {
+		startChainVals = append(startChainVals,
+			StartChainValidator{
+				Id:         val,
+				Stake:      uint(100000000 * (idx + 1)),
+				Allocation: 10000000000,
+			})
+		valBalance[val] = 10000000000
+	}
+
+	// Start the chain
+	step = Step{
+		Action: StartConsumerChainAction{
+			ConsumerChain: ChainID(consumerName),
+			ProviderChain: ChainID("provi"),
+			Validators:    startChainVals,
+		},
+		State: State{
+			ChainID(consumerName): ChainState{
+				ValBalances: &valBalance,
+			},
+		},
+	}
+	s = append(s, step)
+
+	// Establish IBC connection
+	steps := []Step{
```

Comment:
I would add steps that transition a chain from opt-in to topN via a gov proposal.

Reply:
we have that tested already in stepsModifyChain() in `steps_partial_set_security()`
see https://github.com/cosmos/interchain-security/blob/50bbbfedeb82ee2f826569cf03bfb04d5e49e60b/tests/e2e/steps_partial_set_security.go#L2148